xkafka/consumer.go

```diff
@@ -77,57 +84,135 @@ func (c *Consumer) Start(ctx context.Context) error {
 
 	c.handler = c.concatMiddlewares(c.handler)
 
-	st := stream.New().WithMaxGoroutines(c.config.concurrency)
-	errChan := make(chan error, 1)
+	if c.config.concurrency > 1 {
+		return c.runAsync(ctx)
+	}
+
+	return c.runSequential(ctx)
+}
+
+func (c *Consumer) runSequential(ctx context.Context) (err error) {
+	defer func() {
+		if uerr := c.unsubscribe(); uerr != nil {
+			err = errors.Join(err, uerr)
+		}
+	}()
 
 	for {
 		select {
 		case <-ctx.Done():
-			st.Wait()
+			return err
+		default:
+			km, err := c.kafka.ReadMessage(c.config.pollTimeout)
+			if err != nil {
+				var kerr kafka.Error
+				if errors.As(err, &kerr) && kerr.Code() == kafka.ErrTimedOut {
+					continue
+				}
+
+				if ferr := c.config.errorHandler(err); ferr != nil {
+					err = ferr
+
+					return err
+				}
+			}
+
+			msg := newMessage(c.name, km)
+
+			err = c.handler.Handle(ctx, msg)
+			if ferr := c.config.errorHandler(err); ferr != nil {
+				err = ferr
+
+				return err
+			}
+
+			err = c.storeMessage(msg)
+			if err != nil {
+				return err
+			}
+		}
+	}
+}
+
+func (c *Consumer) runAsync(ctx context.Context) error {
+	st := stream.New().WithMaxGoroutines(c.config.concurrency)
+	ctx, cancel := context.WithCancelCause(ctx)
 
-			return c.unsubscribe()
-		case err := <-errChan:
+	for {
+		select {
+		case <-ctx.Done():
 			st.Wait()
 
 			uerr := c.unsubscribe()
-			if uerr != nil {
-				// TODO: use multierror
-				return errors.Wrap(err, uerr.Error())
+
+			err := context.Cause(ctx)
+			if err != nil && errors.Is(err, context.Canceled) {
+				return uerr
 			}
 
-			return err
+			return errors.Join(err, uerr)
 		default:
 			km, err := c.kafka.ReadMessage(c.config.pollTimeout)
-
 			if err != nil {
-				if kerr, ok := err.(kafka.Error); ok && kerr.Code() == kafka.ErrTimedOut {
+				var kerr kafka.Error
+				if errors.As(err, &kerr) && kerr.Code() == kafka.ErrTimedOut {
 					continue
 				}
 
 				if ferr := c.config.errorHandler(err); ferr != nil {
```

Comment:
Similar to [comment](https://github.com/gojekfarm/xtools/pull/26/files#r1428501491)

```go
if ferr := c.config.errorHandler(err); ferr != nil {
	cancel(ferr)
}

continue
```